### PR TITLE
Fixed jupyter problem with RmqThreadCommunicator

### LINKED ISSUE #52: `RmqThreadCommunicator` does not work with recent version of Jupyter
Starting from `notebook==6.0.0`, the `RmqThreadCommunicator` cannot be used in a Jupyter notebook because both applications will try to use the same event loop. This leads to the following exception:
```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-1-4798024621a1> in <module>
      1 import kiwipy
      2 
----> 3 comm = kiwipy.connect('amqp://127.0.0.1')
      4 
      5 

~/code/aiida/env/dev/kiwipy/kiwipy/communicate.py in connect(uri, **kwargs)
      8     if uri.startswith('amqp'):
      9         from . import rmq  # Avoid circular ref: pylint: disable=import-outside-toplevel
---> 10         return rmq.connect(connection_params=uri, **kwargs)
     11 
     12     raise ValueError("Uknown communicator uri '{}'".format(uri))

~/code/aiida/env/dev/kiwipy/kiwipy/rmq/threadcomms.py in connect(connection_params, connection_factory, loop, message_exchange, task_exchange, task_queue, task_prefetch_size, task_prefetch_count, encoder, decoder, testing_mode)
    332                                          encoder=encoder,
    333                                          decoder=decoder,
--> 334                                          testing_mode=testing_mode)

~/code/aiida/env/dev/kiwipy/kiwipy/rmq/threadcomms.py in connect(cls, connection_params, connection_factory, loop, message_exchange, task_exchange, task_queue, task_prefetch_size, task_prefetch_count, encoder, decoder, testing_mode, async_task_timeout)
     56 
     57         # Run the loop to create the connection
---> 58         connection = loop.run_until_complete(connection_factory(**connection_params))
     59         comm = cls(connection,
     60                    message_exchange=message_exchange,

/usr/lib/python3.7/asyncio/base_events.py in run_until_complete(self, future)
    561         """
    562         self._check_closed()
--> 563         self._check_runnung()
    564 
    565         new_task = not futures.isfuture(future)

/usr/lib/python3.7/asyncio/base_events.py in _check_runnung(self)
    524         if events._get_running_loop() is not None:
    525             raise RuntimeError(
--> 526                 'Cannot run the event loop while another loop is running')
    527 
    528     def run_forever(self):

RuntimeError: Cannot run the event loop while another loop is running

```

### PR DESCRIPTION
**Pull request by muhrin**
Fixes #52 

Changed connect() so that now it pipes the arguments through to the
class which will spin up the event loop in a separate thread when
start() is called and at that point connect to RMQ.  This way, all the
loop stuff is being done on the separate thread the jupyter is happy.

This necessitated changing the constructor signature for both
RmqCommunicator and RmqThreadCommunicator, however the respective
connect() function signatures remain unchanged.

Added an async_connect() that creates an RmqCommunicator mirroring the
threaded version.  Updated the tests to use this.

Also made it possible to use an environmental variable to change the
RabbitMQ URI because, e.g., gitlab CI doesn't use 'localhost'.

Added jupyter notebook test - this used to fail because of the loop
issues using RmqThreadCommunicator.  It needs to be ran interactively
though.

### CODE REVIEW

**Comment by muhrin** [comment]
Just going to mention @ltalirz  and @unkcpz here.  If you guys remember our discussion regarding kiwi (and therefore AiiDA) now working with `tornado` versions `>=5`.  This fixes it.

File: test/rmq/jupyter-test.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    return "Thanks!"
+
+
+comm.add_rpc_subscriber(rpc_call, 'tester')
+
+response = comm.rpc_send('tester', 'hello!')
+
+# %%
+
+print(response.result())
```
Comment (sphuber) [inline test/rmq/jupyter-test.py:19]: Do we maybe want to turn this into an assert on the actual expected response?

File: test/rmq/jupyter-test.ipynb [hunk 2]
```diff
@@ -0,0 +1,167 @@
+{
```
Comment (sphuber) [inline test/rmq/jupyter-test.ipynb:1]: Is this somehow magically tested through the CI?

File: test/rmq/jupyter-test.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
+# %%
+
+import kiwipy
```
Comment (sphuber) [inline test/rmq/jupyter-test.py:3]: Is there a particular reason why you have this as a free-standing script and not as a test function?

File: docs/source/conf.py [hunk 4]
```diff
@@ -88,6 +88,7 @@
 #
 # html_theme_options = {}
 html_theme_options = {
+    'analytics_id': 'UA-17296547-4',
```
Comment (sphuber) [inline docs/source/conf.py:91]: unrelated change

File: kiwipy/rmq/communicator.py [hunk 5]
```diff
@@ -262,69 +263,81 @@ async def _send_response(self, reply_to, correlation_id, response):
 
 class RmqCommunicator:
     """
-    A publisher and subscriber using aio_pika and an asyncio event loop
+    A publisher and subscriber using aio_pika and an asyncio event loop.
+
+    The communicator takes an :class:`aio_pika.Connection` as there can be multiple communicators
+    that share a single connection.
```
Comment (sphuber) [inline kiwipy/rmq/communicator.py:269]: Is this correct? I only see `connection_params` and `connection_factory` in constructor args, none of which the type seems to be `aio_pika.Connection`.

File: kiwipy/rmq/communicator.py [hunk 6]
```diff
@@ -262,69 +263,81 @@ async def _send_response(self, reply_to, correlation_id, response):
...
+
+    _connection = None
+    _message_subscriber = None
+    _message_publisher = None
+    _default_task_queue = None  # type: Optional[tasks.RmqTaskQueue]
+
+    def __init__(
+        self,
+        # Connection parameters
+        connection_params: Union[str, dict] = None,
```
Comment (sphuber) [inline kiwipy/rmq/communicator.py:282]: Does it make sense to have the `connection_params` optional? Do all factories, or at least the default one, have sensible defaults that will lead to a useful connection?

File: kiwipy/rmq/communicator.py [hunk 7]
```diff
@@ -334,24 +347,79 @@ async def __aexit__(self, exc_type, exc_val, exc_tb):
...
     @property
     def loop(self):
         return self._connection.loop
 
+    def get_default_task_queue(self) -> Optional[tasks.RmqTaskQueue]:
+        if not self.connected():
+            return None
+
+        if self._default_task_queue is None:
+            self._default_task_queue = tasks.RmqTaskQueue(self._connection,
```
Comment (sphuber) [inline kiwipy/rmq/communicator.py:361]: Should these be connected explicitly when instantiated, like is done in the `connect` method or is this done lazily somehow?

If I understand the logic correctly, the idea is that you always first have to call `connect` which will call these methods, which will construct the task queue, sub and pub, and then calls connect on them. Is there a possibility that if `connect` successfully opens the connection, but than fails or gets interrupted at line 408. Calling `get_message_subscriber` will then happily return the subscriber, but it won't actually be connected. Should this method just lazily do the connection? Something like this:
```python
    def get_default_task_queue(self) -> Optional[tasks.RmqTaskQueue]:
        if not self.connected():
            return None

        if self._default_task_queue is None:
            self._default_task_queue = tasks.RmqTaskQueue(self._connection,
                                                         exchange_name=self._task_exchange,
                                                          queue_name=self._task_queue,
                                                          decoder=self._decoder,
                                                          encoder=self._encoder,
                                                          prefetch_size=self._task_prefetch_size,
                                                          prefetch_count=self._task_prefetch_count,
                                                          testing_mode=self._testing_mode)
        if self._default_task_queue.is_closed:
            self._default_task_queue.connect()

         return self._default_task_queue
```

File: test/rmq/jupyter-test.py (hunk 1 above)
Comment (muhrin) [reply test/rmq/jupyter-test.py:19]: I created this script and the corresponding ipy notebook because they both do the same thing but the jupyter notebook used to fail because of the event loop issues.  I'm not sure if/how these could be executed manually but I don't see the effort/reward ratio being big enough to warrant setting that up given that this specific bug is unlikely to reappear in this form again.  It is good to have them around though so they can be manually invoked just in case.

File: docs/source/conf.py (hunk 4 above)
Comment (muhrin) [reply docs/source/conf.py:91]: Indeed - I didn't see the worth in creating a PR for this but if insist I can do it.

File: kiwipy/rmq/communicator.py (hunk 6 above)
Comment (muhrin) [reply kiwipy/rmq/communicator.py:282]: This is porting in the behaviour of `aio_pika` where all (I think there are only two) of the factories default to "amqp://guest:guest@127.0.0.1/" [1].  I like the behaviour which is why I kept it.

[1] https://aio-pika.readthedocs.io/en/latest/apidoc.html#aio_pika.connect_robust

File: kiwipy/rmq/communicator.py (hunk 7 above)
Comment (muhrin) [reply kiwipy/rmq/communicator.py:361]: I'm not sure I quiet understand the flow of the connection failure you outlined but let me sum up my thinking on this and maybe it'll become clearer.

Basically, for the communicator classes the user has to explicitly instantiate _and_ call the `connect` member function but the preferred way to do this is, of course, the free (factory) function which does both for you.  I don't think the communicator should be lazily connecting if the user forgot because this is a resource expensive and error prone operation and should be done explicitly.

Is there a flow that you think we should put more error checking around?

File: kiwipy/rmq/communicator.py (hunk 5 above)
Comment (muhrin) [reply kiwipy/rmq/communicator.py:269]: Thanks!  This _was_ the case, but no longer.  I'll push a correction.

File: docs/source/conf.py (hunk 4 above)
Comment (sphuber) [reply docs/source/conf.py:91]: ok

File: kiwipy/rmq/communicator.py (hunk 5 above)
Comment (muhrin) [reply kiwipy/rmq/communicator.py]: Ok, done.

**Comment by muhrin** [comment]
Ok, good from my side.  Feel free to have another gander of when you have a chance.

File: kiwipy/rmq/communicator.py (hunk 7 above)
Comment (sphuber) [reply kiwipy/rmq/communicator.py:359]: This will be addressed by PR #53
